Add SVG support (lunasvg)

Lpsd opened the pull request:
I've implemented [lunasvg](https://github.com/sammycage/lunasvg), which gives us the ability to create SVG documents, edit them on the fly, and also load SVG files (from path or raw).

## API (for initial release)

```lua
svg svgCreate(int width, int height [, string pathOrRawData, function callback ])
xmlnode svgGetDocumentXML(svg svgElement)
bool svgSetDocumentXML(svg svgElement, xmlnode svgDocument [, function callback ])
int, int svgGetSize(svg svgElement)
bool svgSetSize(svg svgElement, int width, int height [, function callback ])
```

`callback` is for when SVG has been updated, invoked when texture is available (if update successful).

Example usage
```lua
local square = svgCreate(300, 300)
local squareXML = svgGetDocumentXML(square)

-- Add a rect svg node
local rect = xmlCreateChild(squareXML, "rect")
xmlNodeSetAttribute(rect, "x", "0")
xmlNodeSetAttribute(rect, "y", "0")
xmlNodeSetAttribute(rect, "width", "100%")
xmlNodeSetAttribute(rect, "height", "100%")
xmlNodeSetAttribute(rect, "fill", "red")

-- Apply our changes
svgSetDocumentXML(square, squareXML)

-- Create a square with blank document
local clone = svgCreate(300, 300)

-- Set the fill to blue
xmlNodeSetAttribute(rect, "fill", "blue")

-- Apply xml to blank svg document
svgSetDocumentXML(clone, squareXML)

addEventHandler("onClientRender", root, function()
    dxDrawImage(500, 150, 300, 300, square, 0, 0, 0, tocolor(255, 255, 255), false)
    dxDrawImage(900, 350, 300, 300, clone, 0, 0, 0, tocolor(255, 255, 255), false)
end)
```
```lua
local monkey = svgCreate(500, 500, "monkey.svg")

addEventHandler("onClientRender", root, function()
    dxDrawImage(0, 0, 500, 500, monkey, 0, 0, 0, tocolor(255, 255, 255), false)
end)
```
<details>
<summary>Preview of the first example</summary>

![image](https://user-images.githubusercontent.com/40902730/105227443-3c7d6580-5b59-11eb-86e1-9ed904ab8bad.png)
</details>
<details>
<summary>Preview of the second example</summary>

![image](https://user-images.githubusercontent.com/40902730/105227630-73537b80-5b59-11eb-900c-f9a19dc77263.png)
</details>

The `svg` element is also derived from CClientTexture, meaning it works with pretty much all dx* functions, including `dxGetTexturePixels`, `dxSetTexturePixels`, `dxSetPixelColor`, etc - direct manipulation is therefore possible, e.g:

```lua
local svg = svgCreate(300, 300, "svg.svg")

-- Bind to clear the SVG texture
bindKey("r", "down", function()
    local pixels = dxGetTexturePixels(svg)
    local width, height = svgGetSize(svg)
    
    for i = 0, width do
        for j = 0, height do
            dxSetPixelColor(pixels, i, j, 0, 0, 0, 0)
        end
    end

    dxSetTexturePixels(svg, pixels)
end)
```

(Resolves #1300)

Lpsd committed:
Add lunasvg vendor

Lpsd committed:
Add CVectorGraphic & CLuaVectorGraphicDefs

Lpsd committed:
Clean up/extend implementation and add CRenderItem.VectorGraphic

Lpsd committed:
Add CClientVectorGraphicDisplay

Lpsd committed:
Add UpdateTexture method to CClientVectorGraphic

Lpsd committed:
Remove unnecessary forward declaration in CClientManager

Lpsd committed:
Small fix to SVGDocument initialization in CClientVectorGraphic

Lpsd committed:
Remove unintended m_uiStatsWebBrowserCount increase in CClientRenderElementManager

Lpsd committed:
Update CreateDocument test in CClientVectorGraphic

Lpsd committed:
Change CClientVectorGraphic to inherit CClientRenderElement instead of CClientTexture

Lpsd committed:
Fix issue with render item initialization (causing crash)

Also changed some minor things with the CreateDocument test code

Lpsd committed:
Update premake and use dllexport

Lpsd committed:
Revert to old arg parser

I'm not sure how to return my custom element CClientVectorGraphic (it's the same as CClientWebBrowser)

Lpsd committed:
Fix small issues

Lpsd committed:
Add .svg file reading and fix continuous re-drawing of same texture causing lag

Pirulax commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,110 @@
+    m_pManager = pManager;
+
+    m_pRenderItem = pVectorGraphicItem; // This is cast to CClientRenderItem
+    m_pVectorGraphicItem = pVectorGraphicItem; // Keep this as CVectorGraphicItem so we don't have to cast everywhere later on...
```
`static_cast` is compile time, that is, it takes no additional time run time.

Lpsd replied:
regardless of this I think it's just cleaner to store it here, so others looking at the code in future know it's intrinsic to the CClientVectorGraphic class

Lpsd committed:
Add svgAddRect function

Lpsd committed:
Add BGRA (default) support to lunasvg

Lpsd committed:
Fix AddRect and implement AddCircle

Lpsd committed:
Add forgotten ColorFormat check (fixes color)

Lpsd committed:
Return debugscript error on fail to load SVG file/data, and fix possible crashes

Lpsd committed:
Addendum to previous commit

prnxdev commented:
Nice!!

I think there is no need for new function `svgCreate` for files. I think that `DxCreateTexture` should work with `.svg` too the same way it works for `.png` or `.jpg`. Also OOP support would be nice :)

Lpsd commented:
Integrating into the DX related code would not be good for the overall structure of the API.

`svgCreate` returns a texture element, `CClientVectorGraphic`, which can be passed directly to `dxDrawImage`.

`CClientVectorGraphic` encapsulates all of the SVG related data, the D3D9 render item and the graphics display class. None of the other dx* functions follow this convention, and in reality they're two completely separate constructs. It would be a mess to implement both features side by side.

Not to mention, we're going to have a LOT of functions with the svg* prefix, to manipulate the SVG document and load data. Somehow, `dxCreateSVGDocument` doesn't seem right to me?

Have a look into the code, to get a better understand of the implementation.

Lpsd committed:
Move drawing stuff from CClientVectorGraphic to CClientVectorGraphicDisplay

Lpsd commented:
As briefly discussed on Discord, I think the initial plan should be to roll this out with a few basic functions:

```lua
texture/svg svgCreate(width, height [, string pathOrRawData ])
xmlnode svgGetDocumentXML(svg svgElement)
bool svgSetDocumentXML(xmlnode xmlData)
```

You can create a blank SVG document, supply a file path or read from raw (including xmlnode) data. You can also get the SVG document back as an xmlnode, use the existing xml implementation to edit the xml data, and set it back on the document.

The reason for introducing this simple behaviour at first, is because a proper API will take a lot of time and thought, and will probably be quite an addition to the code. Introducing the basics in this PR, and then the API stuff later, is easier on the reviewer(s).

We can always deprecate `svgGet/SetDocumentXML`  at a later date, once a better API has been introduced. 

Thoughts?

sbx320 commented on Client/core/Graphics/CRenderItem.VectorGraphic.cpp:
```diff
@@ -0,0 +1,141 @@
+
+#include "StdInc.h"
+
+////////////////////////////////////////////////////////////////
```
We don't really use these massive doxygen-like comments anymore

sbx320 commented on Client/core/Graphics/CRenderItemManager.cpp:
```diff
@@ -143,6 +143,29 @@ CTextureItem* CRenderItemManager::CreateTexture(const SString& strFullFilePath,
+////////////////////////////////////////////////////////////////
+CVectorGraphicItem* CRenderItemManager::CreateVectorGraphic(uint width, uint height)
+{
+    CVectorGraphicItem* pVectorItem = new CVectorGraphicItem();
```
Should use a `std::unique_ptr` here

sbx320 commented on Client/core/Graphics/CRenderItemManager.cpp:
```diff
@@ -143,6 +143,29 @@ CTextureItem* CRenderItemManager::CreateTexture(const SString& strFullFilePath,
+    if (!pVectorItem->IsValid())
+    {
+        SAFE_RELEASE(pVectorItem);
+        return NULL;
```
`nullptr`

sbx320 commented on Client/mods/deathmatch/StdInc.h:
```diff
@@ -104,6 +104,7 @@
 #include <CScriptFile.h>
 #include <CWeaponNames.h>
 #include <CVehicleNames.h>
+#include <CClientVectorGraphic.h>
```
There's no need to add this to the precompiled header. Since it is rarely used throughout the project, it should only be included where it is needed. 

The same also applies to the existing code, but fixing that is a lot more work than doing it correctly for new stuff

sbx320 commented on Client/mods/deathmatch/StdInc.h:
```diff
@@ -138,6 +139,7 @@
 #include <luadefs/CLuaTaskDefs.h>
 #include <luadefs/CLuaTeamDefs.h>
 #include <luadefs/CLuaTimerDefs.h>
+#include <luadefs/CLuaVectorGraphicDefs.h>
```
Same here

sbx320 commented on Client/mods/deathmatch/logic/CClientEntity.h:
```diff
@@ -78,6 +78,7 @@ enum eClientEntityType
     CCLIENTSEARCHLIGHT,
     CCLIENTIFP,
     CCLIENTUNKNOWN,
+    CCLIENTVECTORGRAPHIC,
```
I'm not sure if we use `CCLIENTUNKNOWN` somewhere as a size indicator for the enum. Should be checked

sbx320 commented on Client/mods/deathmatch/logic/CClientRenderElementManager.cpp:
```diff
@@ -242,6 +242,31 @@ CClientWebBrowser* CClientRenderElementManager::CreateWebBrowser(uint uiSizeX, u
+
+    // Check create worked
+    if (!pVectorGraphicItem)
+        return NULL;
```
nullptr

sbx320 commented on Client/mods/deathmatch/logic/CClientRenderElementManager.cpp:
```diff
@@ -242,6 +242,31 @@ CClientWebBrowser* CClientRenderElementManager::CreateWebBrowser(uint uiSizeX, u
+        return NULL;
+
+    // Create the element
+    CClientVectorGraphic* pVectorGraphicElement = new CClientVectorGraphic(m_pClientManager, INVALID_ELEMENT_ID, pVectorGraphicItem);
```
Ideally this would use a `unique_ptr`, might not be that easy to add though due to the existing map.

Lpsd replied:
As discussed in Discord, this should probably be done in a separate PR as it would require changes to existing code, shared with quite a few other classes.

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,130 @@
+    m_pRenderItem = pVectorGraphicItem; // This is cast to CClientRenderItem
+    m_pVectorGraphicItem = pVectorGraphicItem; // Keep this as CVectorGraphicItem so we don't have to cast everywhere later on...
+
+    m_pVectorGraphicDisplay = new CClientVectorGraphicDisplay(m_pManager->GetDisplayManager(), this);
```
Should be `std::unique_ptr` or just a member

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,130 @@
+    if (m_pDocument)
+        return;
+
+    m_pDocument = new SVGDocument();
```
Should be `std::unique_ptr`

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,130 @@
+
+    SVGElement* rootElement = m_pDocument->rootElement();
+
+    // There has to be a better way to do this (:
```
We can use `std::visit` like:

```
    struct ToString {
        std::string operator()(float val) { return std::to_string(val); }
        std::string operator()(const std::string& val) { return val; }
    };
    std::variant<std::string, int> var = ...;
    auto str = std::visit(ToString{}, var);
```
which makes it a bit nicer. Could also wrap this in a custom "variant-to-string" function.

Lpsd replied:
For the initial release of this, I'm not planning to have the SVGAddRect or SVGAddCircle functions, so it's not needed for now. We can discuss a proper implementation of this later.

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,130 @@
+        (ry.index() == 0) ? std::to_string(std::get<float>(ry)).c_str() : std::get<string>(ry).c_str(),
+        std::to_string(pathLength).c_str(),
+        fill.c_str(),
+        std::to_string(++m_iShapeCount).c_str());
```
I'd pull the `++` into a separate line. Right now it is very easy to miss.

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,130 @@
+    return false;
+}
+
+const bool CClientVectorGraphic::IsDestroyed()
```
Returning a const value (as in: not a ref, not a ptr) is pointless since you can store it in a `bool` anyway

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,55 @@
+
+    std::variant<bool, int> AddRect(std::variant<float, std::string> x, std::variant<float, std::string> y, std::variant<float, std::string> width, std::variant<float, std::string> height, std::variant<float, std::string> rx, std::variant<float, std::string> ry, float pathLength, std::string fill);
+
+    std::variant<bool, int> CClientVectorGraphic::AddCircle(std::variant<float, std::string> cx, std::variant<float, std::string> cy, float radius, float pathLength, std::string fill);
```
The `CClientVectorGraphic::` does not belong here

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,55 @@
+    bool LoadFromFile(std::string strFilePath);
+
+    void       Destroy();
+    const bool IsDestroyed();
```
see above regarding `const bool`. The method should be const, not the return value.

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,55 @@
+    SVGDocument*        GetSVGDocument() { return m_pDocument; }
+    CVectorGraphicItem* GetRenderItem() { return m_pVectorGraphicItem; }
+
+    const bool IsDisplayCleared() { return m_pVectorGraphicDisplay->IsCleared(); }
```
see above regarding `const bool`. The method should be const, not the return value.

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,55 @@
+    void       Destroy();
+    const bool IsDestroyed();
+
+    SVGDocument*        GetSVGDocument() { return m_pDocument; }
```
These two should also be `const`

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.cpp:
```diff
@@ -0,0 +1,167 @@
+    UpdateTexture();
+}
+
+CClientVectorGraphicDisplay::~CClientVectorGraphicDisplay()
```
You can just implement this by saying `= default` in the header

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.cpp:
```diff
@@ -0,0 +1,167 @@
+
+    SVGDocument* svgDocument = m_pVectorGraphic->GetSVGDocument();
+
+    if(!m_pVectorGraphic->GetSVGDocument())
```
This should've been `svgDocument == nullptr`

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.cpp:
```diff
@@ -0,0 +1,167 @@
+    uint width = pVectorGraphicItem->m_uiSizeX;
+    uint height = pVectorGraphicItem->m_uiSizeY;
+
+    Bitmap bitmap = svgDocument->renderToBitmap(width, height, 96.0);
```
`96.0`?

Lpsd replied:
96 dpi

sbx320 replied:
Would be good to add a comment for that (or even a constexpr constant)

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.h:
```diff
@@ -0,0 +1,65 @@
+    friend class CClientDisplayManager;
+
+public:
+    CClientVectorGraphicDisplay(CClientDisplayManager* pDisplayManager, CClientVectorGraphic* pVectorGraphic, int ID = 0xFFFFFFFF);
```
`ID = 0xFFFFFFFF` should have some form of constant name

Lpsd replied:
I've set it to `DISPLAY_VECTORGRAPHIC` - it made the most sense.

https://github.com/multitheftauto/mtasa-blue/blob/master/Client/mods/deathmatch/logic/CClientTextDisplay.h will need the `ID` updating to `DISPLAY_TEXT` in a separate PR.

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,133 @@
+void CLuaVectorGraphicDefs::LoadFunctions()
+{
+   constexpr static const std::pair<const char*, lua_CFunction> functions[]{
+        {"svgCreate", SVGCreate},
```
Is there a reason why this doesn't use the ArgumentParser yet?

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,133 @@
+                        if (!pVectorGraphic->LoadFromFile(strPath))
+                        {
+                            pVectorGraphic->Destroy();
+                            delete pVectorGraphic;
```
This delete looks very out of place

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,133 @@
+    return 1;
+}
+
+std::variant<bool, int> CLuaVectorGraphicDefs::SVGAddRect(CClientVectorGraphic* pVectorGraphic, std::variant<float, std::string> x, std::variant<float, std::string> y,
```
You might want to add some general type `using` declaration for `std::variant<float, std::string>` to keep this a bit more readable

Lpsd replied:
Related to https://github.com/multitheftauto/mtasa-blue/pull/2026#discussion_r560978524

Lpsd committed:
Convert svgCreate to new arg parser, and extend implementation with CClientVectorGraphic::LoadFromData

Lpsd committed:
Throw exceptions in CLuaVectorGraphicDefs instead of returning nullptr

Lpsd committed:
Clean up file header comments

Lpsd committed:
Remove unnecessary includes from mods/deathmatch/StdInc.h

Lpsd committed:
Implement suggested unique_ptr changes, add constexpr DPI and change NULL -> nullptr

Lpsd committed:
Move CCLIENTUNKNOWN to end of eClientEntityType enum

Lpsd committed:
Add vector graphic counter to CClientPerfStat.LuaMemory.cpp

It seems the code which actually increases/decreases this value is in another commit, 6ef21a350620b9b761a9bbaa065c89316fd454d3

Lpsd committed:
Add check for svg in CClientEntity::GetTypeID

Lpsd committed:
Change CClientVectorGraphicDisplay ID to DISPLAY_VECTORGRAPHIC

This will need changing in CClientTextDisplay too

Lpsd commented:
I've updated the first post with the proposed API for initial release, and updated the examples

Lpsd committed:
Remove AddRect and AddCircle related methods

Lpsd committed:
Add ToString to CXMLNodeImpl

Lpsd committed:
Fix issue with return logic

Lpsd committed:
Return empty string instead of nullptr

sbx320 commented on Client/core/Graphics/CRenderItemManager.cpp:
```diff
@@ -135,14 +135,37 @@ CTextureItem* CRenderItemManager::CreateTexture(const SString& strFullFilePath,
+////////////////////////////////////////////////////////////////
+std::unique_ptr<CVectorGraphicItem> CRenderItemManager::CreateVectorGraphic(uint width, uint height)
+{
+    std::unique_ptr<CVectorGraphicItem> pVectorItem(new CVectorGraphicItem());
```
Use `std::make_unique` for this kind of stuff

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,82 @@
+    m_pManager = pManager;
+
+    m_pVectorGraphicItem = std::unique_ptr<CVectorGraphicItem>(std::move(pVectorGraphicItem));
+    m_pVectorGraphicDisplay = std::unique_ptr<CClientVectorGraphicDisplay>(new CClientVectorGraphicDisplay(m_pManager->GetDisplayManager(), this));
```
`std::make_unique`

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,82 @@
+    m_pResource = nullptr;
+    m_pManager = pManager;
+
+    m_pVectorGraphicItem = std::unique_ptr<CVectorGraphicItem>(std::move(pVectorGraphicItem));
```
Just do `m_pVectorGraphicItem = std::move(pVectorGraphicItem)`. No need for a new `unique_ptr`

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,82 @@
+    if (m_pDocument)
+        return;
+
+    m_pDocument = std::unique_ptr<SVGDocument>(new SVGDocument());
```
`std::make_unique`

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,53 @@
+
+using namespace lunasvg;
+
+constexpr float DPI() { return 96.0; }
```
There are constexpr variables! :) 

`constexpr float DPI = 96.0f;`

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,53 @@
+    CClientVectorGraphic(CClientManager* pManager, ElementID ID, std::unique_ptr<CVectorGraphicItem> pVectorGraphicItem);
+    ~CClientVectorGraphic() = default;
+
+    CResource* GetResource() { return m_pResource; }
```
can also be const

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,53 @@
+    std::unique_ptr<CVectorGraphicItem>               m_pVectorGraphicItem;
+    std::unique_ptr<CClientVectorGraphicDisplay>      m_pVectorGraphicDisplay;
+
+    int m_iShapeCount;
```
Is this used anywhere?

Lpsd replied:
It was, but not anymore, forgot to remove.

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,53 @@
+
+    int m_iShapeCount;
+    
+    bool m_bIsDestroyed;
```
This needs to be initialized

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.cpp:
```diff
@@ -0,0 +1,164 @@
+    D3DLOCKED_RECT LockedRect;
+    surface->LockRect(&LockedRect, nullptr, 0);
+
+    device->ColorFill(surface, NULL, D3DCOLOR_ARGB(0, 0, 0, 0));
```
`nullptr`

sbx320 commented on Shared/XML/CXMLNodeImpl.cpp:
```diff
@@ -553,3 +553,19 @@ void CXMLNodeImpl::SetCommentText(const char* szCommentText, bool bLeadingBlankL
+    if (pDocument == nullptr)
+        return nullptr;
+
+    std::unique_ptr<TiXmlPrinter> pPrinter(new TiXmlPrinter());
```
You don't need to heap allocate it here. Just do 
```
TiXmlPrinter printer;
```

Lpsd replied:
Fixed in #2031 - can you review/merge that PR and then I can merge master into this branch again?

Lpsd committed:
Merge branch 'master' into lunasvg

Lpsd committed:
Merge branch 'xmltostring' into lunasvg

Lpsd committed:
Remove heap allocation

Lpsd committed:
Change unique_ptr to use make_unique & add Set/GetSVGDocument

Also changed constexpr function to variable and changed a NULL to nullptr

Lpsd committed:
Fix issue

Lpsd committed:
Merge branch 'xmltostring' into lunasvg

Lpsd committed:
Add svgSetDocumentXML and svgGetDocumentXML

Using old arg parser, I can't work out how to handle CXMLNode with the new one?

Lpsd commented:
I've added `svgDocumentGetXML` & `svgDocumentSetXML` and updated the main post with examples & previews of how it works.

Lpsd commented:
Converted from draft pull request. Requesting final review.

(Note: relies on https://github.com/multitheftauto/mtasa-blue/pull/2031)

Lpsd commented:
actually, I found a crash in release mode. I'll fix this first :)

Lpsd committed:
Fix crash caused by uneccessary unique_ptr member variable

For some reason I was storing a CVectorGraphicItem object as a member (unique_ptr) on CClientVectorGraphic, but this is not needed, it is already stored as m_pRenderItem (CRenderItem inherited class), and a manager controls it (which was resulting in a crash).

Lpsd commented:
Fixed crash, see 859537ae9

Also, some testing I did...

![image](https://user-images.githubusercontent.com/40902730/105542431-3d48ff80-5cf1-11eb-93ef-e8be2d2ab1f1.png)

```lua
local iter = 1000

startBenchmark("Run simple task " .. iter .. " times")

for i = 1, iter do
    local circle = svgCreate(300, 300)

    local xml = svgGetDocumentXML(circle)

    -- Add a circle svg node
    local node = xmlCreateChild(xml, "circle")
    xmlNodeSetAttribute(node, "cx", "150")
    xmlNodeSetAttribute(node, "cy", "150")
    xmlNodeSetAttribute(node, "r", "50%")
    xmlNodeSetAttribute(node, "fill", "#0fc0fc")

    -- Apply our changes
    svgSetDocumentXML(circle, xml)
end

finishBenchmark()
```

I tried just running `svgCreate(300, 300)` too

![image](https://user-images.githubusercontent.com/40902730/105543035-7b462380-5cf1-11eb-8966-0aa7716e8dc8.png)

Not bad!

Pirulax commented:
Probably most of the time is spent reading the data from the disk.

Lpsd commented:
No data ever gets written to disk, and it's only read from disk if you load an .svg file with `svgCreate`

sbx320 committed:
Fix lua::Push not working on non-element entities
Fixes #1658

sbx320 commented on Client/core/Graphics/CRenderItemManager.cpp:
```diff
@@ -135,14 +135,40 @@ CTextureItem* CRenderItemManager::CreateTexture(const SString& strFullFilePath,
+CVectorGraphicItem* CRenderItemManager::CreateVectorGraphic(uint width, uint height)
+{
+    if (!CanCreateRenderItem(CVectorGraphicItem::GetClassId()))
+        return NULL;
```
nullptr

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.cpp:
```diff
@@ -0,0 +1,164 @@
+#include "CClientVectorGraphicDisplay.h"
+#include "CClientVectorGraphic.h"
+
+using std::list;
```
this looks unused

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,144 @@
+{
+   constexpr static const std::pair<const char*, lua_CFunction> functions[]{
+        {"svgCreate", ArgumentParser<SVGCreate>},
+        {"svgGetDocumentXML", SVGGetDocumentXML},
```
Use argument parser (with #2042)

sbx320 commented on Shared/XML/CXMLNodeImpl.cpp:
```diff
@@ -553,3 +553,14 @@ void CXMLNodeImpl::SetCommentText(const char* szCommentText, bool bLeadingBlankL
+    if (m_pNode->Accept(&printer))
+        return std::string(printer.CStr());
+
+    return std::string("");
```
`return {}`

Lpsd committed:
Update returns in CXMLNodeImpl::ToString()

Lpsd committed:
Merge branch 'xmltostring' into lunasvg

Lpsd committed:
Merge branch 'master' into lunasvg

Lpsd committed:
Merge branch 'bugfix/argparser-entities' of https://github.com/sbx320/mtasa-blue into lunasvg

Lpsd committed:
Merge branch 'master' into lunasvg

Lpsd committed:
Fix minor issues

Lpsd committed:
Add some CClientVectorGraphic defs in lua logic

Lpsd committed:
Use new arg parser for svgSet/GetDocumentXML

Lpsd commented:
Suggested changes have been implemented, tested and working well.

Lpsd committed:
Merge branch 'master' into lunasvg

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,115 @@
+
+CClientVectorGraphic* CLuaVectorGraphicDefs::SVGCreate(lua_State* luaVM, CVector2D size, std::optional<std::string> pathOrRawData)
+{
+    if (size.fX < 0 || size.fY < 0 || size.fX == 0 || size.fY == 0)
```
```suggestion
    if (size.fX <= 0 || size.fY <= 0)
```

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,115 @@
+        {
+            // Try and load from the filepath or svg data (if provided)
+            std::string strPath;
+            if (CResourceManager::ParseResourcePathInput(pathOrRawData.value(), pParentResource, &strPath) && FileExists(strPath))
```
You could always check if the string begins with a `<` after skipping whitespace and take the SVG-data branch in that case. With the `&& FileExists(strPath)` condition here and with no readable file in the filesystem, you would present the scripter an error with `Unable to load SVG data`.

botder commented on Client/mods/deathmatch/logic/CClientDisplayManager.cpp:
```diff
@@ -88,7 +88,7 @@ void CClientDisplayManager::DoPulse()
     {
         CClientDisplay* pObject = *iter;
-        if (pObject->IsExpired())
+        if (pObject->GetType() == DISPLAY_TEXT && pObject->IsExpired())
```
Why did you change this code?

Lpsd replied:
This was for testing, I forgot to revert the change.

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,53 @@
+#include <svgdocument.h>
+#include "CClientVectorGraphicDisplay.h"
+
+using namespace lunasvg;
```
Never use `using namespace` in header files.

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,53 @@
+
+using namespace lunasvg;
+
+constexpr float DPI = 96.0;
```
Make it a class constant of `CClientVectorGraphic`. Don't introduce a global constant.

Lpsd replied:
Added to the class

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,53 @@
+
+#pragma once
+
+#include <svgdocument.h>
```
Since you only store a pointer to the SVG document, you could avoid pulling in the library header in this header file.

Lpsd replied:
I think I need this header. 

When I took out the `#include <svgdocument.h>` (replaced with forward declaration for SVGDocument), `GetSVGDocumentXML` complained (pointer to incomplete class type is not allowed)... 

Anyway, I've done this -> 11d0768ce

sbx320 replied:
You could move the inline definition to the soruce file to avoid the header inclusion.

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.h:
```diff
@@ -0,0 +1,64 @@
+    unsigned long m_ulFormat;
+    unsigned char m_ucShadowAlpha;
+
+    static float m_fGlobalScale;
```
Do you even use it? What is it for?

Lpsd replied:
Copied from another class which inherits `CClientDisplay` (which I used as reference) - probably not needed so will remove.

Lpsd replied:
Actually, I will probably remove this class entirely and integrate the logic directly into `CClientVectorGraphic` instead. 

I based this on a similar class purely so I could use existing render pipelines, but looking at it now it doesn't seem necessary. I'll take a proper look later on.

Lpsd replied:
Decided to just remove unnecessary definitions here.

Lpsd committed:
Update size check in SVGCreate

Co-authored-by: Marek Kulik <me@botder.com>

Lpsd committed:
Merge branch 'master' into lunasvg

Lpsd committed:
Update premake5.lua (add vendor/lunasvg)

Lpsd committed:
Remove "using namespace" from header and define DPI under class

Lpsd committed:
Revert test code in CClientDisplayManager

Lpsd committed:
Add using namespace lunasvg where necessary and fix DPI references

Lpsd committed:
Remove unnecessary definitions in CClientVectorGraphicDisplay

Lpsd committed:
Improve filepath/raw-data handling in SVGCreate

sbx320 commented on premake5.lua:
```diff
@@ -137,6 +137,7 @@ workspace "MTASA"
 		include "vendor/pthreads"
 		include "vendor/libspeex"
 		include "vendor/detours"
+    	include "vendor/lunasvg"
```
Indentation error

sbx320 commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.cpp:
```diff
@@ -0,0 +1,125 @@
+
+CClientVectorGraphicDisplay::CClientVectorGraphicDisplay(CClientDisplayManager* pDisplayManager, CClientVectorGraphic* pVectorGraphic, int ID) : CClientDisplay(pDisplayManager, ID)
+{
+    // Init
```
That comment seems a bit pointless

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+
+    lua_classfunction(luaVM, "create", "svgCreate");
+    lua_classfunction(luaVM, "getDocumentXML", "svgGetDocumentXML");
+    lua_classfunction(luaVM, "SetDocumentXML", "svgSetDocumentXML");
```
typo: `SetDocumentXML` should be `setDocumentXML`.

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+    CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine(luaVM);
+    CResource* pParentResource = pLuaMain->GetResource();
+
+    CClientVectorGraphic* pVectorGraphic = g_pClientGame->GetManager()->GetRenderElementManager()->CreateVectorGraphic((int)size.fX, (int)size.fY);
```
`static_cast`

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+
+        if (pathOrRawData.has_value())
+        {
+            std::string strRawData = pathOrRawData.value().c_str();
```
`c_str()` seems pointless, `pathOrRawData.value()` is already a `std::string`

patrikjuvonen replied:
@Lpsd This change still seems to be pending.

Lpsd replied:
It's resolved already

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+        if (pathOrRawData.has_value())
+        {
+            std::string strRawData = pathOrRawData.value().c_str();
+            std::size_t szCharPos = strRawData.find_first_not_of(" \t", 0);
```
Shouldn't this also include newlines? In that case `isblank` may be a better option.

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+
+    CClientVectorGraphic* pVectorGraphic = g_pClientGame->GetManager()->GetRenderElementManager()->CreateVectorGraphic((int)size.fX, (int)size.fY);
+
+    if (pVectorGraphic)
```
If you invert the logic here, we can save an indentation level

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+
+CXMLNode* CLuaVectorGraphicDefs::SVGGetDocumentXML(lua_State* luaVM, CClientVectorGraphic* pVectorGraphic)
+{
+    CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine(luaVM);
```
Maybe we should add CLuaMain* to the argument parser to avoid this recurring pattern

Lpsd replied:
Agreed, we should open another issue for that.

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+        CXMLNode* rootNode = pLuaMain->ParseString(strXmlContent.c_str());
+
+        if (rootNode && rootNode->IsValid())
+        {
```
Braces around if but not else looks weird

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,126 @@
+    throw std::invalid_argument("Unable to get SVG XML document");
+}
+
+bool CLuaVectorGraphicDefs::SVGSetDocumentXML(lua_State* luaVM, CClientVectorGraphic* pVectorGraphic, CXMLNode* pXMLNode)
```
This never returns false, might as well return void instead

Lpsd committed:
Remove svgdocument.h include from CClientVectorGraphic.h and move inline definition

Lpsd committed:
Add required header file to CClientVectorGraphicDisplay.cpp and remove unnecessary comment

Lpsd committed:
Fix logic for SVGCreate and formatting in CLuaVectorGraphicDefs

Lpsd committed:
Fix premake5.lua tab spacing

Lpsd committed:
Remove unnecessary c_str()

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+        std::string strRawData = pathOrRawData.value();
+        std::size_t szCharPos = strRawData.find_first_not_of(" \t\n\r\f\v", 0);
+
+        const char charAuto = strRawData.at(szCharPos);
```
`std::string::at` is a bounds-checking and throwing function. Please compare `szCharPos` to `std::string::npos` to check if the end of the string has been reached

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+        {
+            std::string strPath;
+            if (CResourceManager::ParseResourcePathInput(pathOrRawData.value(), pParentResource, &strPath) && FileExists(strPath))
+            {
```
You are not checking if the current resource has permission to access an another resource.

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+    CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine(luaVM);
+    CResource* pParentResource = pLuaMain->GetResource();
+
+    CClientVectorGraphic* pVectorGraphic = g_pClientGame->GetManager()->GetRenderElementManager()->CreateVectorGraphic(static_cast<int>(size.fX), static_cast<int>(size.fY));
```
What happens to this object when this function fails later below?

Pirulax replied:
How'd that happen?

Lpsd replied:
forgot to call `Destroy` on the object before throwing

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.h:
```diff
@@ -0,0 +1,41 @@
+#include "CClientDisplay.h"
+#include "CClientDisplayManager.h"
+
+class CClientVectorGraphicDisplay : public CClientDisplay
```
```suggestion
class CClientVectorGraphicDisplay final : public CClientDisplay
```

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.cpp:
```diff
@@ -0,0 +1,125 @@
+#include "StdInc.h"
+#include "CClientVectorGraphicDisplay.h"
+#include "CClientVectorGraphic.h"
+#include "svgdocument.h"
```
```suggestion
#include <svgdocument.h>
```

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphic.h:
```diff
@@ -0,0 +1,56 @@
+    class SVGDocument;
+}
+
+class CClientVectorGraphic : public CClientTexture
```
```suggestion
class CClientVectorGraphic final : public CClientTexture
```

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphicDisplay.h:
```diff
@@ -0,0 +1,41 @@
+    void ClearTexture();
+
+    const bool IsCleared() { return m_bIsCleared; }
+    const bool HasUpdated() { return m_bHasUpdated; }
```
```suggestion
    bool IsCleared() const { return m_bIsCleared; }
    bool HasUpdated() const { return m_bHasUpdated; }
```

Lpsd committed:
Merge branch 'master' into lunasvg

Pirulax commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+
+    if (pathOrRawData.has_value())
+    {
+        std::string strRawData = pathOrRawData.value();
```
could use `std::string&` or `*pathOrRawData*` or `pathOrRawData->`

Lpsd replied:
not entirely sure what you mean with this

Pirulax commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+
+CXMLNode* CLuaVectorGraphicDefs::SVGGetDocumentXML(lua_State* luaVM, CClientVectorGraphic* pVectorGraphic)
+{
+    CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine(luaVM);
```
Use `lua_getownercluamain`

Pirulax commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+        std::string strXmlContent = pVectorGraphic->GetSVGDocumentXML();
+
+        CXMLNode* rootNode = pLuaMain->ParseString(strXmlContent.c_str());
+
```
unnecessary copy.
```suggestion
        CXMLNode* rootNode = pLuaMain->ParseString(pVectorGraphic->GetSVGDocumentXML().c_str());
```

Pirulax commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+        if (rootNode && rootNode->IsValid())
+            return rootNode;
+        else
+            throw std::invalid_argument("Unable to get SVG XML document");
```
```suggestion
```
The throw below is the same, isn't it? If so, this can be omitted.

Pirulax commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,120 @@
+{
+    std::string strXML = pXMLNode->ToString();
+
+    if (pVectorGraphic->LoadFromData(strXML))
```
```suggestion
    if (pVectorGraphic->LoadFromData(pXMLNode->ToString()))
```
This is perfectly readable.

Lpsd committed:
Merge branch 'master' into lunasvg

Lpsd committed:
Make vector classes "final"

Lpsd committed:
include optimiziation

Lpsd committed:
out-of-bounds check in SVGCreate and some cleanup

Lpsd commented:
Requested changes made, just a couple of comments from @Pirulax above which I'm unsure about (seems fine imo).

Lpsd commented:
It seems the lunasvg project has been heavily refactored since this was implemented, so I'm in the process of upgrading - some big changes to the API.

Lpsd commented:
seems like we can't really upgrade unless we make heavy changes to the latest vendor version ourselves - a lot of classes and implementations have been restricted / made private in the refactor, and we need access to those. 

One thing that's been restricted (unless I'm completely blind) is the ability to get/set and manipulate the XML for the document directly.

kind of a shame - do we find a new SVG lib or keep this one (the current version in this PR) knowing we would have to maintain it ourselves?

edit: there are a few small changes I could make to the newest vendor in order to make this work, otherwise we would have to load XML content via lunasvg and then load it again on the MTA side to keep track of the document data, which isn't great.

I'll have another stab at it shortly

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,118 @@
+
+        if (rootNode && rootNode->IsValid())
+            return rootNode;
+    }
```
```suggestion
    CLuaMain& luaMain = lua_getownercluamain(luaVM);
    CXMLNode* rootNode = luaMain.ParseString(pVectorGraphic->GetSVGDocumentXML().c_str());

    if (rootNode && rootNode->IsValid())
        return rootNode;
```

botder commented:
> One thing that's been restricted (unless I'm completely blind) is the ability to get/set and manipulate the XML for the document directly.

That has been removed in the 2.0.0 release. This is the commit before the change: https://github.com/sammycage/lunasvg/tree/2d6d80e835a607a3e4e148620b9f376f479a2d31

sammycage commented:
> kind of a shame - do we find a new SVG lib or keep this one (the current version in this PR) knowing we would have to maintain it ourselves?

@Lpsd version 2.x.x focuses more on rendering. Since svg is xml, you can use a library like [tinyxml](https://github.com/multitheftauto/mtasa-blue/tree/c58c3678f263d810f61eb5aecba48e6c4b4dce0c/vendor/cegui-0.4.0-custom/src/tinyxml) to do the manipulation and creation, then you can send it to lunasvg for rendering.

https://github.com/sammycage/lunasvg/issues/28#issuecomment-790667351

Lpsd commented:
@sammycage We are already using tinyxml in this implementation, so manipulating the XML isn't really the problem (we can add back any missing implementation ourselves), it's retrieving the current document data.

Is there a possibility of bringing back the `toString` method, or something similar? Otherwise we have to load the file twice - once on lunasvg side through `loadFromFile`, then again on our side to keep track of the XML data we're manipulating. 

edit: I guess we could also just load the file once on our side and then use `loadFromData` instead.... huh, why didn't I think of that earlier 😅

sammycage commented:
> edit: I guess we could also just load the file once on our side and then use loadFromData instead.... huh, why didn't I think of that earlier

What about the `toString` method?

Lpsd commented:
> What about the `toString` method?

If we load the file and store the data on our side, then use `loadFromData` exclusively (instead of `loadFromFile`) then I guess we won't need the old `toString` method. It completely passed me by that we can do it this way :)

sammycage commented:
> > What about the `toString` method?
> 
> If we load the file and store the data on our side, then use `loadFromData` exclusively (instead of `loadFromFile`) then I guess we won't need the old `toString` method. It completely passed me by that we can do it this way :)

Nice idea 💯💯. This is exactly what i'm thinking 😊😊😊.

sammycage commented:
@Lpsd is this necessary?

https://github.com/multitheftauto/mtasa-blue/blob/83af268b982ba9a3ae0dc211cc9a954af9c18d45/Client/mods/deathmatch/logic/CClientVectorGraphic.cpp#L45-L49

Lpsd committed:
Update lunasvg vendor to latest (v2.2.0)

Lpsd committed:
Rewrite code to use updated lunasvg API

Lpsd committed:
Fix memory leak in CClientVectorGraphic::LoadFromData

Lpsd commented:
> @Lpsd is this necessary?

No, this is outdated following the latest 3 commits. I have updated our implementation to use the latest version of lunasvg.

This is ready for another review by the way.

botder commented on vendor/lunasvg/premake5.lua:
```diff
@@ -0,0 +1,40 @@
+  flags { "MultiProcessorCompile" }
+
+configuration "vs2019 and Release"
+  flags { "LinkTimeOptimization" } 
```
Is this really necessary?

Lpsd replied:
no, removed

botder commented on vendor/lunasvg/premake5.lua:
```diff
@@ -0,0 +1,40 @@
+project "lunasvg"
```
Please indent this premake5.lua with tabs to be consistent

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,132 @@
+        throw std::invalid_argument("Unable to set SVG XML document");
+
+    return;
+}
```
```suggestion
bool CLuaVectorGraphicDefs::SVGSetDocumentXML(CClientVectorGraphic* pVectorGraphic, CXMLNode* pXMLNode)
{
    return pVectorGraphic->LoadFromData(pXMLNode->ToString());
}
```

Lpsd replied:
Why has the throw been removed in your suggestion?

botder replied:
I would rather prefer a boolean than a Lua error.

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,132 @@
+    return pVectorGraphic;
+}
+
+CXMLNode* CLuaVectorGraphicDefs::SVGGetDocumentXML(lua_State* luaVM, CClientVectorGraphic* pVectorGraphic)
```
```suggestion
CXMLNode* CLuaVectorGraphicDefs::SVGGetDocumentXML(CClientVectorGraphic* pVectorGraphic)
```

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.h:
```diff
@@ -0,0 +1,24 @@
+
+    static CClientVectorGraphic* SVGCreate(lua_State* luaVM, CVector2D size, std::optional<std::string> pathOrRawData);
+    static CXMLNode*             SVGGetDocumentXML(lua_State* luaVM, CClientVectorGraphic* pVectorGraphic);
+    static void                  SVGSetDocumentXML(lua_State* luaVM, CClientVectorGraphic* pVectorGraphic, CXMLNode* pXMLNode);
```
```suggestion
    static CXMLNode*             SVGGetDocumentXML(CClientVectorGraphic* pVectorGraphic);
    static bool                  SVGSetDocumentXML(CClientVectorGraphic* pVectorGraphic, CXMLNode* pXMLNode);
```

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,132 @@
+        {
+            if (!pVectorGraphic->LoadFromData(pathOrRawData.value()))
+            {
+                pVectorGraphic->Destroy();
```
You don't destroy the object itself `CClientVectorGraphic* pVectorGraphic`

Lpsd replied:
Removed `Destroy` method, it's now handled via `CClientVectorGraphic::Unlink` (called via destructor, as well as by element manager)

Lpsd committed:
Merge branch 'master' into lunasvg

Lpsd committed:
Update premake5.lua indentation (tabs)

Lpsd committed:
Fix vector graphic element destruction in SVGCreate and SVGSetDocumentXML return value (non-throw)

Lpsd committed:
Add extra checks to be safe in CClientVectorGraphicDisplay

Lpsd committed:
Fix CClientVectorGraphic destructor (use Unlink) and remove unnecessary definitions

Lpsd commented:
As a note for the latest commit (14df0ae), `m_pXMLDocument` is a member of type `std::unique_ptr<SXMLString>` (returned by `CXMLImpl::ParseString`)

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,61 @@
+    m_pDocument = nullptr;
+    m_pXMLDocument = nullptr;
+
+    m_pResource = nullptr;
```
Nitpicking here, but could you please do that in the header file instead?

botder commented on Client/mods/deathmatch/logic/CClientVectorGraphic.cpp:
```diff
@@ -0,0 +1,61 @@
+CXMLNode* CClientVectorGraphic::GetSVGDocumentXML() const
+{
+    return m_pXMLDocument;
+}
```
Move this method definition to the header file

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,129 @@
+        std::string& strRawData = pathOrRawData.value();
+        std::size_t szCharPos = strRawData.find_first_not_of(" \t\n\r\f\v", 0);
+
+        char charAuto;
```
```suggestion
        char charAuto = 0;
```
Variable wasn't initialized.

botder commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,129 @@
+        {
+            if (!pVectorGraphic->LoadFromData(pathOrRawData.value()))
+            {
+                delete pVectorGraphic;
```
Are you sure about this? Since you set the parent/owner resource for the vector graphic above (line 52 & line 55) - aren't they going to use access or even delete it themselves? (You might also have this issue on line 88, 100 and 107)

Lpsd committed:
Move stuff from CClientVectorGraphic.cpp into header

Lpsd committed:
Update CClientVectorGraphic destruction in CLuaVectorGraphicDefs

botder reviewed (approved):
It looks good now, but needs testing.
Anyway, I am not sure if we want to use hard errors in certain cases (see `throw`s) because that makes failure handling much harder in script code - we should keep hard errors for invalid parameters, but maybe not for other cases.

Lpsd commented:
To be honest I think we can even allow the CClientVectorGraphic object to live on after failing to load data in `SVGCreate` (instead of deleting the object and throwing). We should just have a warning in debugscript instead, the user can still use that created SVG object to try load more data on the document - it's up to him to destroy it (or it'll be destroyed when the resource stops anyway).

For example, if you fail to load a webpage in a browser, it won't destroy the browser object, you can just try loading another URL. Makes more sense tbh

Lpsd committed:
Merge branch 'master' into lunasvg

Lpsd committed:
Remove throws in Lua definitions

Return nil and displays debug error instead

Lpsd committed:
Fix SXMLString ownership and slight refactor / renaming

Lpsd committed:
Fix build (not returning a value in SVGCreate)

oops

Lpsd commented:
Fixed issues addressed in the svg-graphics channel on Discord from testing. Stuff has changed a bit in CClientVectorGraphic (e13871e) so could do with another check. Tested a lot of scenarios and no crashes.

Regarding my last comment it would be weird for the SVG loading to fail, but for the object to still "live on" - even if you could set the XML data again, you wouldn't be able to tell it had actually failed (at script level, you'd expect nil/false to catch it).

For that reason it now warns in debug console and returns nil instead - no throws aside from when invalid arguments are provided (as per @botder's suggestion above).

patrikjuvonen commented on vendor/lunasvg/README.md:
```diff
@@ -0,0 +1,87 @@
+[![Releases](https://img.shields.io/badge/Version-2.2.0-orange.svg)](https://github.com/sammycage/lunasvg/releases)
```
Nit: Even though no technical changes, can we update lunasvg to v2.3.0 (https://github.com/sammycage/lunasvg/releases/tag/v2.3.0)

Lpsd replied:
Done

Lpsd committed:
Avoid "points" overflow on outline creation in plutovg-rle.c

Lpsd committed:
Add "contours" to overflow check in plutovg-rle.c

Lpsd committed:
Revert plutovg-rle.c "overflow" checks

Lpsd committed:
Use memmove instead of memcpy to avoid overflow

patrikjuvonen commented on vendor/lunasvg/3rdparty/software/sw_ft_stroker.c:
```diff
@@ -662,8 +662,6 @@ static void ft_stroke_border_export(SW_FT_StrokeBorder border,
 
     outline->n_points = (short)(outline->n_points + border->num_points);
-
-    assert(SW_FT_Outline_Check(outline) == 0);
```
Is this is an unwanted change in the vendor library source code? Vendor code at https://github.com/sammycage/lunasvg/blob/b1b9a1ce0add9cd963f5ee33cb38088e98bc3ac7/3rdparty/software/sw_ft_stroker.c#L665-L666

Lpsd replied:
no, if the `memmove` now "fails" due to buffer overlap (which we don't need to care about) then this assertion will happen - we don't want that

patrikjuvonen replied:
Perhaps we should instead comment it out and add "Changed by MTA" as an extra comment, so it doesn't get accidentally added during an upgrade in the future.

patrikjuvonen commented on vendor/lunasvg/3rdparty/software/sw_ft_stroker.c:
```diff
@@ -0,0 +1,1937 @@
+{
+    /* copy point locations */
+    memmove(outline->points + outline->n_points, border->points,
+           border->num_points * sizeof(SW_FT_Vector));
```
Let's make a distinction here that we've changed the original vendor library source code, so that we don't lose them during future upgrades.
```suggestion
    /** Changed by MTA */
    /* move point locations */
    memmove(outline->points + outline->n_points, border->points,
           border->num_points * sizeof(SW_FT_Vector));
    /** */
```

Lpsd replied:
Added comments for both the memcpy -> memmove change and assertion removal

Lpsd committed:
Add svgGetSize (returns width, height)

Lpsd committed:
Add svgSetSize

Lpsd committed:
Add min-size check to svgSetSize

patrikjuvonen reviewed (commented):
First of all great work on this pull request @Lpsd! :tada:

I think this pull request is soon ready to be merged.

I tested this pull request by using the InstallFiles artifact from commit [8d50b1fae4ec9350611ec734ea0f223fe59c54cd](https://github.com/multitheftauto/mtasa-blue/pull/2026/commits/8d50b1fae4ec9350611ec734ea0f223fe59c54cd).

I used your two examples, as well as a new example to test linear gradient and radial gradient. All looks good.

I also tested two libraries:
- [Font Awesome 5.15.4](https://github.com/FortAwesome/Font-Awesome/releases/tag/5.15.4): All 1611 SVGs loaded and rendered (eyed through, looks good)
- [Material Design Icons 4.0.0](https://github.com/google/material-design-icons/releases/tag/4.0.0): All 7681 SVGs loaded and rendered (eyed through, looks good)

I also tested the SVG 1.2 Tiny Test Suite from 12 September 2008:
- [SVG 1.2 Tiny Test Suite (12 September 2008)](https://www.w3.org/Graphics/SVG/WG/wiki/Test_Suite_Overview#SVG_1.2_Tiny_Test_Suite:_12_September_2008): 491 out of 495 SVGs loaded and rendered (61 passed tests, 434 failed tests)

The W3 SVG test suite is of course quite old and extensive and I guess would be better directed at lunasvg. To my surprise gradients in the W3 test suite did not work with lunasvg, but again, it is soon a 12-year-old test suite, so could be anything. It includes tests for animations, scripting and events for example, which we do not plan on supporting anyway. So that test is sort of useless in our use case. Was still fun to run it and see how much passes.

My full SVG test suite (drive.google.com because I couldn't upload it here on GitHub for some reason...?): https://drive.google.com/file/d/1K77MS7XFvGnPL6NkWedIH-6uOI8ZP337/view?usp=sharing. It is large, so it may take a while to load on the client.

API looks good at this stage!

Overall again great work.

I will go through the code in a new review later and after any changes you make, so I will not approve this just yet.

Lpsd committed:
Add back missing file load check in SVGCreate

Lpsd committed:
Update lunasvg to v2.3.0

Lpsd committed:
Add buffer overflap/overflow fix with comments

Lpsd committed:
Add optional callback parameters to Lua definitions

Lpsd commented:
Added optional callback args to `svgCreate`, `svgSetSize` and `svgSetDocumentXML`. Useful for knowing exactly when the texture is ready after update.

Here's an example, binds for each "test" are displayed on screen:  [svg.zip](https://github.com/multitheftauto/mtasa-blue/files/7193261/svg.zip)

Lpsd committed:
Remove unnecessary defs in async tasks

Lpsd committed:
Fix callback crashes due to invalid reference

Lpsd committed:
Add 4096 size limit (dx9 limitation)

Lpsd committed:
Addition to 9678bd

Lpsd committed:
Fix colors in lunasvg (rgba bug?)

sammycage commented on vendor/lunasvg/source/lunasvg.cpp:
```diff
@@ -195,7 +195,9 @@ void Document::render(Bitmap bitmap, const Matrix& matrix, std::uint32_t backgro
     root->render(state);
-    state.canvas->rgba();
+
+    /** Changed by MTA */
```
PlutoVG image format is ARGB32 Premultiplied not BGRA

sammycage replied:
ARGB Premultiplied -> BGRA Non-Premultiplied

```cpp
auto width = state.canvas->width();
auto height = state.canvas->height();
auto stride = state.canvas->stride();
auto data = state.canvas->data();
for(int y = 0;y < height;y++)
{
    auto pixels = reinterpret_cast<uint32_t*>(data + stride * y);
    for(int x = 0;x < width;x++)
    {
        auto pixel = pixels[x];
        auto a = (pixel >> 24) & 0xFF;
        if(a == 0)
            continue;

        auto r = (pixel >> 16) & 0xFF;
        auto g = (pixel >> 8) & 0xFF;
        auto b = (pixel >> 0) & 0xFF;
        if(a != 255)
        {
            r = (r * 255) / a;
            g = (g * 255) / a;
            b = (b * 255) / a;
        }

        pixels[x] = (a << 24) | (r << 16) | (g << 8) | b;
    }
}
```

Lpsd replied:
aha, I see - thanks!

Lpsd committed:
Use "clamp" texture address to avoid edge artifacts

patrikjuvonen committed:
Merge branch 'master' into lunasvg

patrikjuvonen committed:
Merge branch 'master' into lunasvg

patrikjuvonen committed:
Run clang-format

patrikjuvonen committed:
Merge branch 'master' into lunasvg

patrikjuvonen committed:
Use https based mtasa url

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,318 @@
+                    m_pScriptDebugging->LogCustom(luaVM, "Unable to load SVG data (check for XML syntax errors)");
+                    return false;
+                }
+            }
```
The code is currently duplicated twice, once for the callback and once for the non-callback. I would like this to be refactored into a e.g. single private function call.

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,318 @@
+                            return false;
+                        }
+                    }
+                }
```
The code is currently duplicated twice, once for the callback and once for the non-callback. I would like this to be refactored into a e.g. single private function call.

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,318 @@
+        return true;
+    }
+
+    return pVectorGraphic->SetDocument(pXMLNode);
```
The code is currently duplicated twice, once for the callback and once for the non-callback. I would like this to be refactored into a e.g. single private function call.

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaVectorGraphicDefs.cpp:
```diff
@@ -0,0 +1,318 @@
+
+    pVectorGraphic->GetDisplay()->Update();
+
+    return true;
```
The code is currently duplicated twice, once for the callback and once for the non-callback. I would like this to be refactored into a e.g. single private function call.

patrikjuvonen reviewed (changes requested):
Thanks for the changes! I have now reviewed the code and overall looks good!

I saw a lot of code style formatting issues, are you running clang-format on your editor? We also have a clang-formatter batch script in the utils folder, which helps with fixing formatting issues to be consistent with the rest of the codebase.

Either way to speed up the process I decided to run clang-format on this PR and push the changes here. 

Only one thing bugs me which is that the callback and non-callback code is duplicated, which is not best practice. I would like to see these refactored into a single private function call for example, or something else that doesn't result in duplicating code, thus duplicating work and duplicating amount of potential issues.

The size functions work great, thank you for adding them!

I think once you've refactored the duplicate code, this is ready for a final review and then merge.

patrikjuvonen committed:
Remove github files

patrikjuvonen committed:
Remove nonexistant folder link

patrikjuvonen committed:
Fix typo

Lpsd committed:
Add static methods for callbacks

Lpsd commented:
Thanks for the review - I've split that code into static, private methods as suggested. In some cases it doesn't cut down the amount of code needed but at least it's more consistent for any future changes.

Lpsd committed:
Update formatting

Lpsd commented:
I've had to format the `CLuaShared::GetAsyncTaskScheduler()->PushTask` calls by hand, clang-format makes them really ugly if the task function only contains a single line (like a return). Rest should be fine

patrikjuvonen committed:
Remove old file line from file comment

patrikjuvonen reviewed (approved):
Thank you very much, everything looks good now! 😄
